src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,607 @@
+            result.kvs.insert(result.kvs.end(), list.kvs.begin(),list.kvs.end());
+
+          biter++;
+          if (list.next != "")
+            result.next = list.next;
+          else if (biter != eiter)
+            result.next = biter->get_node_val();
+          else
+            result.next = "";
+
+          return list_ertr::make_ready_future<bool>(false);
+        });
+      });
+    }).safe_then([&result] {
+      return list_ertr::make_ready_future<list_kvs_result_t>(std::move(result));
+    });
+  });
+}
+
+OMapInnerNode::clear_ret
+OMapInnerNode::clear(omap_context_t oc)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  return crimson::do_for_each(iter_begin(), iter_end(), [this, oc] (auto iter) {
+    auto laddr = iter->get_node_key().laddr;
+    return omap_load_extent(oc, this, laddr, get_meta().depth - 1).safe_then(
+      [oc] (auto &&extent) {
+      return extent->clear(oc);
+    }).safe_then([oc, laddr] {
+      return oc.tm.dec_ref(oc.t, laddr);
+    }).safe_then([] (auto ret){
+      return clear_ertr::now();
+    });
+  });
+}
+
+OMapInnerNode::split_children_ret
+OMapInnerNode:: make_split_children(omap_context_t oc, OMapNodeRef pnode)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  return omap_alloc_2extents<OMapInnerNode>(oc, pnode, OMAP_BLOCK_SIZE)
+    .safe_then([this] (auto &&ext_pair) {
+      auto [left, right] = ext_pair;
+      return split_children_ret(
+             split_children_ertr::ready_future_marker{},
+             std::make_tuple(left, right, split_into(*left, *right)));
+  });
+}
+
+OMapInnerNode::full_merge_ret
+OMapInnerNode::make_full_merge(omap_context_t oc, OMapNodeRef right, OMapNodeRef pnode)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  return omap_alloc_extent<OMapInnerNode>(oc, pnode, OMAP_BLOCK_SIZE)
+    .safe_then([this, right] (auto &&replacement) {
+      replacement->merge_from(*this, *right->cast<OMapInnerNode>());
+      return full_merge_ret(
+        full_merge_ertr::ready_future_marker{},
+        std::move(replacement));
+  });
+}
+
+OMapInnerNode::make_balanced_ret
+OMapInnerNode::make_balanced(omap_context_t oc, OMapNodeRef _right,
+                             bool prefer_left, OMapNodeRef pnode)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  ceph_assert(_right->get_type() == type);
+  return omap_alloc_2extents<OMapInnerNode>(oc, pnode, OMAP_BLOCK_SIZE)
+    .safe_then([this, _right, prefer_left] (auto &&replacement_pair){
+      auto [replacement_left, replacement_right] = replacement_pair;
+      auto &right = *_right->cast<OMapInnerNode>();
+      return make_balanced_ret(
+             make_balanced_ertr::ready_future_marker{},
+             std::make_tuple(replacement_left, replacement_right,
+             balance_into_new_nodes(*this, right, prefer_left,
+                                    *replacement_left, *replacement_right)));
+  });
+}
+
+OMapInnerNode::checking_parent_ret
```

Comment:
See my other comment about split and merge.  I think this method exists because splitting and merging during the descent isn't really a valid simplification here.  I suspect that doing the splits and merges starting at the leaf will cause this to be replaced with something simpler.

If you disagree, it would help to have a precise explanation of what this method does.